core/iwasm/interpreter/wasm_interp_fast.c

```diff
@@ -38,9 +38,33 @@ typedef float64 CellType_F64;
 #endif
 
 #if WASM_ENABLE_SHARED_HEAP != 0
-#define app_addr_in_shared_heap(app_addr, bytes)        \
-    (shared_heap && (app_addr) >= shared_heap_start_off \
-     && (app_addr) <= shared_heap_end_off - bytes + 1)
+/* TODO: add code for 64 bit version when memory64 is enabled for fast-interp */
+#define get_shared_heap_start_off(shared_heap) (shared_heap->start_off_mem32)
+/* Check whether the app addr in the last visited shared heap, if not, check the
+ * shared heap chain to find which(if any) shared heap the app addr in, and
+ * update the last visited shared heap info if found. */
+#define app_addr_in_shared_heap(app_addr, bytes)                               \
+    (shared_heap && (app_addr) >= shared_heap_start_off                        \
+     && (app_addr) <= shared_heap_end_off - bytes + 1)                         \
+        || ({                                                                  \
+               bool in_chain = false;                                          \
+               WASMSharedHeap *cur;                                            \
+               uint64 cur_shared_heap_start_off, cur_shared_heap_end_off;      \
+               for (cur = shared_heap; cur; cur = cur->chain_next) {           \
+                   cur_shared_heap_start_off = get_shared_heap_start_off(cur); \
+                   cur_shared_heap_end_off =                                   \
+                       cur_shared_heap_start_off - 1 + cur->size;              \
+                   if ((app_addr) >= cur_shared_heap_start_off                 \
+                       && (app_addr) <= cur_shared_heap_end_off - bytes + 1) { \
+                       shared_heap_start_off = cur_shared_heap_start_off;      \
+                       shared_heap_end_off = cur_shared_heap_end_off;          \
+                       shared_heap_base_addr = cur->base_addr;                 \
+                       in_chain = true;                                        \
+                       break;                                                  \
+                   }                                                           \
+               }                                                               \
+               in_chain;                                                       \
+           })
```

Comment:
Is there a way we could combine `app_addr_in_shared_heap()` from both *interp_classic* and *interp_fast* into a common location, such as *wasm_memory.h*, to eliminate code duplication?